controllers/users_config_map_controller.go

```diff
@@ -0,0 +1,274 @@
+
+func IsUsersConfigMap(object client.Object) bool {
+	if configMap, ok := object.(*corev1.ConfigMap); ok {
+		if configMap.Name == additionalUsersConfigMapName {
+			return true
+		}
+	}
+	return false
+}
+
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=imagerepositories,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;update;patch
+
+func (r *ConfigMapReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithName("ConfigMap")
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// fetch the config map instance
+	configMap := &corev1.ConfigMap{}
+	if err := r.Client.Get(ctx, req.NamespacedName, configMap); err != nil {
+		if errors.IsNotFound(err) {
+			// The object is deleted, nothing to do
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "failed to get config map", l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+		controllerutil.AddFinalizer(configMap, ConfigMapFinalizer)
+
+		if err := r.Client.Update(ctx, configMap); err != nil {
+			log.Error(err, "failed to add finalizer to config map", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+			return ctrl.Result{}, err
+		}
+		log.Info("finalizer added to configmap")
+		// leave all other steps to reconcile from update action
+		return ctrl.Result{}, nil
+	}
+
+	teamName := fmt.Sprintf("%s-team", configMap.Namespace)
+	removeTeam := false
+
+	if !configMap.DeletionTimestamp.IsZero() {
+		removeTeam = true
+		log.Info("Config map with additional users was removed, will delete team", "TeamName", teamName)
+
+		if controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+			controllerutil.RemoveFinalizer(configMap, ConfigMapFinalizer)
+			if err := r.Client.Update(ctx, configMap); err != nil {
+				log.Error(err, "failed to remove config map finalizer", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+				return ctrl.Result{}, err
+			}
+			log.Info("finalizer removed from config map", l.Action, l.ActionDelete)
+		}
+	}
+
+	var additionalUsers []string
+	if !removeTeam {
+		// get additional users from config map
+		additionalUsersStr, usersExist := configMap.Data[additionalUsersConfigMapKey]
+		if !usersExist {
+			log.Info("Config map with additional users doesn't have the key", "ConfigMapName", additionalUsersConfigMapName, "ConfigMapKey", additionalUsersConfigMapKey, l.Action, l.ActionView)
+			removeTeam = true
+		} else {
+			additionalUsers = strings.Fields(strings.TrimSpace(additionalUsersStr))
+			log.Info("Additional users configured in config map", "AdditionalUsers", additionalUsers)
+		}
+	}
+
+	// reread quay token
+	r.QuayClient = r.BuildQuayClient(log)
+
+	// remove team if config map was removed, renamed, or doesn't contain any users
```

Comment:
mmm but I am doing that

at beginning I am checking if it has finalizer, if not I'll add it and stop and update will trigger new reconcile

here I am also checking if 'quay.io' key is missing from the config map, and in that case I am removing team as well

ImageRepository controller doesn't deal with finalizer at all only ensures Team (which is just safety net, since I want to add permissions to team)

in normal situation

new controller will create team